service/src/prometheus_metrics.rs

```diff
@@ -111,22 +162,47 @@ where
 	}
 }
 
-impl<Wallet> MetricsHandler<Wallet>
+impl<Wallet, Sidechain> MetricsHandler<Wallet, Sidechain>
 where
-	Wallet: EnclaveAccountInfo + Send + Sync,
+	Wallet: ParentchainAccountInfo + Send + Sync,
+	Sidechain: ParentchainIntegriteeSidechainInfo + Send + Sync,
 {
-	pub fn new(enclave_wallet: Arc<Wallet>) -> Self {
-		MetricsHandler { enclave_wallet }
+	pub fn new(wallets: Vec<Arc<Wallet>>, sidechain: Arc<Sidechain>) -> Self {
+		MetricsHandler { wallets, sidechain }
 	}
 
-	async fn update_metrics(&self) {
-		match self.enclave_wallet.free_balance() {
-			Ok(b) => {
-				ENCLAVE_ACCOUNT_FREE_BALANCE.set(b as i64);
-			},
-			Err(e) => {
-				error!("Failed to fetch free balance metric, value will not be updated: {:?}", e);
-			},
+	async fn update_account_metrics(&self) {
+		for wallet in &self.wallets {
+			if let (Ok(balance), Ok(parentchain_id), Ok(account_and_role), Ok(decimals)) = (
+				wallet.free_balance(),
+				wallet.parentchain_id(),
+				wallet.account_and_role(),
+				wallet.decimals(),
+			) {
+				ACCOUNT_FREE_BALANCE
```

Comment:
So, if only one call fails here, we branch into the else statement. I suppose this might be a bit hard to debug, shouldn't we separate them?

Reply:
indeed